src/blackbox.py

```diff
@@ -0,0 +1,270 @@
+
+
+class ConfigUpdateFailure(WorkloadManagerError):
+    """Custom exception for failed config updates."""
+
+
+class ContainerNotReady(WorkloadManagerError):
+    """Raised when an operation is run that presumes the container being ready.."""
+
+
+class WorkloadManager(Object):
+    """Workload manager for blackbox exporter."""
+
+    _layer_name = _service_name = "blackbox-exporter"
+    _exe_name = "blackbox_exporter"
+
+    def __init__(
+        self,
+        charm,
+        *,
+        container_name: str,
+        port: int,
+        web_external_url: str,
+        config_path: str,
+    ):
+        # Must inherit from ops 'Object' to be able to register events.
+        super().__init__(charm, f"{self.__class__.__name__}-{container_name}")
+
+        self._unit = charm.unit
+
+        self._service_name = self._container_name = container_name
+        self._container = charm.unit.get_container(container_name)
+
+        self.api = BlackboxExporterApi(endpoint_url=charm._external_url)
+
+        self._port = port
+        self._web_external_url = web_external_url
+        self._config_path = config_path
+
+        # turn the container name to a valid Python identifier
+        snake_case_container_name = self._container_name.replace("-", "_")
+        charm.framework.observe(
+            charm.on[snake_case_container_name].pebble_ready, self._on_pebble_ready
+        )
+
+    @property
+    def is_ready(self):
+        """Is the workload ready to be interacted with?"""
+        return self._container.can_connect()
+
+    def _on_pebble_ready(self, _):
+        if version := self._blackbox_exporter_version:
+            self._unit.set_workload_version(version)
+        else:
+            logger.debug(
+                "Cannot set workload version at this time: couldn't get Blackbox Exporter version"
+            )
+
+    @property
+    def _blackbox_exporter_version(self) -> Optional[str]:
+        """Returns the version of Blackbox Exporter.
+
+        Returns:
+            A string equal to the Blackbox Exporter version.
+        """
+        if not self.is_ready:
+            return None
+        version_output, _ = self._container.exec([self._exe_name, "--version"]).wait_output()
+        # Output looks like this:
+        # blackbox_exporter, version 0.24.0 (branch: HEAD, ...)
+        result = re.search(r"version (\d*\.\d*\.\d*)", version_output)
+        if result is None:
+            return result
+        return result.group(1)
+
+    def _blackbox_exporter_layer(self) -> Layer:
+        """Returns Pebble configuration layer for Blackbox Exporter."""
+
+        def _command():
+            """Returns full command line to start Blackbox Exporter."""
+            return (
+                f"{self._exe_name} "
+                f"--config.file={self._config_path} "
+                f"--web.listen-address=:{self._port} "
+                f"--web.external-url={self._web_external_url}"
```

Comment:
This looks too similar to alertmanager. Does bbe need such args?

Reply:
Yes, and the reason why it's very similar because it's developed by Prometheus :D